builder error: generic struct

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**
```
OS: linux, Ubuntu 20.04.3 LTS (WSL 2)
Processor: 4 cpus, 64bit, little endian, AMD Ryzen 3 2200G with Radeon Vega Graphics
CC version: cc (Ubuntu 11.1.0-1ubuntu1~20.04) 11.1.0

getwd: /home/simon/lib/struct_mapper
vmodules: /home/simon/.vmodules
vroot: /home/simon/v
vexe: /home/simon/v/v
vexe mtime: 2021-11-11 22:37:52
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 32b74dd.4b42dca

Git version: git version 2.25.1
Git vroot status: weekly.2021.45-23-g4b42dcad
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de
```

**What did you do?**
```v
fn print_field_values<T>(s &T) {
	$for field in T.fields {
		println(s.$(field.name))
	}
}

fn main() {
	bar := Foo{
		name: "Simon"
		password: "simon1234"
		email: "simon@gmail.com"
		age: 15
	}
	print_field_values<Foo>(bar)
}
```

**What did you expect to see?**
```
Simon
simon1234
simon@gmail.com
15
```
 
**What did you see instead?**
```
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c: In function ‘main__print_field_values_T_main__Foo’:
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:12017:48: error: incompatible type for argument 1 of ‘main__Foo_str’
12017 |                         println(main__Foo_str(s->name));
      |                                               ~^~~~~~
      |                                                |
      |                                                string
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:2112:39: note: expected ‘main__Foo’ but argument is of type ‘string’
 2112 | static string main__Foo_str(main__Foo it) { return indent_main__Foo_str(it, 0);}
      |                             ~~~~~~~~~~^~
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:12026:48: error: incompatible type for argument 1 of ‘main__Foo_str’
12026 |                         println(main__Foo_str(s->password));
      |                                               ~^~~~~~~~~~
      |                                                |
      |                                                string
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:2112:39: note: expected ‘main__Foo’ but argument is of type ‘string’
 2112 | static string main__Foo_str(main__Foo it) { return indent_main__Foo_str(it, 0);}
      |                             ~~~~~~~~~~^~
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:12035:48: error: incompatible type for argument 1 of ‘main__Foo_str’
12035 |                         println(main__Foo_str(s->email));
      |                                               ~^~~~~~~
      |                                                |
      |                                                string
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:2112:39: note: expected ‘main__Foo’ but argument is of type ‘string’
 2112 | static string main__Foo_str(main__Foo it) { return indent_main__Foo_str(it, 0);}
      |                             ~~~~~~~~~~^~
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:12044:48: error: incompatible type for argument 1 of ‘main__Foo_str’
12044 |                         println(main__Foo_str(s->age));
      |                                               ~^~~~~
      |                                                |
      |                                                int
/tmp/v_1000/struct_mapper.12266723168755046067.tmp.c:2112:39: note: expected ‘main__Foo’ but argument is of type ‘int’
 2112 | static string main__Foo_str(main__Foo it) { return indent_main__Foo_str(it, 0);}
      |                             ~~~~~~~~~~^~
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -810,6 +810,16 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 		if rec_type.has_flag(.shared_f) {
 			rec_type = rec_type.clear_flag(.shared_f).set_nr_muls(0)
 		}
+		if node.left is ast.ComptimeSelector {
+			if node.left.field_expr is ast.SelectorExpr {
+				if node.left.field_expr.expr is ast.Ident {
+					key_str := '${node.left.field_expr.expr.name}.typ'
+					rec_type = g.comptime_var_type_map[key_str] or { rec_type }
+					g.gen_expr_to_string(node.left, rec_type)
+					return
+				}
+			}
+		}
 		g.get_str_fn(rec_type)
 	} else if node.name == 'free' {
 		mut rec_type := node.receiver_type
@@ -1113,6 +1123,14 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 				g.writeln('; ${c_name(print_method)}($tmp); string_free(&$tmp);')
 			} else {
 				g.write('${c_name(print_method)}(')
+				if expr is ast.ComptimeSelector {
+					if expr.field_expr is ast.SelectorExpr {
+						if expr.field_expr.expr is ast.Ident {
+							key_str := '${expr.field_expr.expr.name}.typ'
+							typ = g.comptime_var_type_map[key_str] or { typ }
+						}
+					}
+				}
 				g.gen_expr_to_string(expr, typ)
 				g.write(')')
 			}

--- a/vlib/v/tests/comptime_for_in_field_selector_test.v
+++ b/vlib/v/tests/comptime_for_in_field_selector_test.v
@@ -1,0 +1,30 @@
+fn print_field_values<T>(s T) {
+	mut value_list := []string{}
+
+	$for field in T.fields {
+		println(s.$(field.name))
+		value_list << s.$(field.name).str()
+	}
+	assert value_list.len == 4
+	assert value_list[0] == 'Simon'
+	assert value_list[1] == 'simon1234'
+	assert value_list[2] == 'simon@gmail.com'
+	assert value_list[3] == '15'
+}
+
+struct Foo {
+	name     string
+	password string
+	email    string
+	age      int
+}
+
+fn test_comptime_for_in_field_selector() {
+	bar := Foo{
+		name: 'Simon'
+		password: 'simon1234'
+		email: 'simon@gmail.com'
+		age: 15
+	}
+	print_field_values<Foo>(bar)
+}

